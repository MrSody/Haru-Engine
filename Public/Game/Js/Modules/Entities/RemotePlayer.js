import Player from './Player.js';

export default class RemotePlayer extends Player {
    constructor (datos) {
        super(datos);
    }

/* ------------------------------ *
    GETTERS
* ------------------------------ */

/* ------------------------------ *
SETTERS
* ------------------------------ */

    setMode (mode) {
        this.mode = mode;
    }

/* ------------------------------ *
    FUNCIONES
* ------------------------------ */

    // Actualiza la posicion del jugador remoto
    posNow (widthMap, heightMap, posWorld) {

        let posIntX = Math.floor(posWorld.x - widthMap),
            posIntY = Math.floor(posWorld.y - heightMap),
            posEndX = Math.floor(posWorld.x + widthMap),
            posEndY = Math.floor(posWorld.y + heightMap);

        if ((posIntX >= this.pos.x || this.pos.x <= posEndX) && (posIntY >= this.pos.y || this.pos.y <= posEndY)) {
            for(let y = 0, intY = posIntY; intY <= posEndY; intY++, y++) {
                for(let x = 0, intX = posIntX; intX <= posEndX; intX++, x++) {

                    if (this.pos.x == intX && this.pos.y == intY) {
                        return {x: x, y: y};
                    }
                }
            }
        }

        return false;
    }

/* ------------------------------ *
    DRAW
* ------------------------------ */

    draw (ctx, HUB, cXnull, cYnull) {
        const tileSize = 32;
        cXnull *= tileSize;
        cYnull *= tileSize;

        this.updateFrame();

        HUB.fillStyle = "#FFF";
        HUB.font = "9pt Minecraftia";
        // Muestra el nombre y el level
        HUB.fillText(this.name +', '+ this.level, cXnull, cYnull - 32);
        // Muestra la vida
        HUB.fillStyle="#000000";
        HUB.fillRect(cXnull, cYnull - 30, 50, 2);
        HUB.fillStyle="#FF0000";
        HUB.fillRect(cXnull, cYnull - 30, 40, 2);

        this.drawMode(ctx, cXnull, cYnull);

        let hitDelta = Date.now() - this.lastHitPointUpdate;

        if (this.hitArray.length > 0 && hitDelta > 50) {
            for (var i = 0, j = this.hitArray.length; i < j; i+=1) {
                if (this.hitArray[i] != null) {
                    this.lastHitPointUpdate = Date.now();
                    this.hitArray[i][1] = Math.round((this.hitArray[i][1]-0.1)*10)/10;

                    if (this.hitArray[i][1] <= 0) {
                        this.hitArray.splice(i,1);
                    }
                }
            }
        } else if (this.hitArray.length == 0) {
            this.gotHit = false;
        }

        if (this.hitArray.length > 0) {
            for (var i = 0, max = this.hitArray.length; i < max; i += 1) {
                ctx.fillStyle = 'rgba(255,0,0,'+this.hitArray[i][1]+')';
                ctx.font = "14pt Minecraftia";
                ctx.fillText(this.hitArray[i][0], this.pos["x"]+12, this.pos["y"]-20);
            }
        }
    }

}