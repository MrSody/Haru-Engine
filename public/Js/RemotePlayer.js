class RemotePlayer extends Player {
    constructor (datos) {
        super(datos);
    }

/* ------------------------------ *
    GETTERS
* ------------------------------ */

/* ------------------------------ *
SETTERS
* ------------------------------ */

/* ------------------------------ *
    FUNCIONES
* ------------------------------ */

// Actualiza la posicion del jugador remoto
posNow (widthMap, heightMap, posWorld) {

    let posIntX = Math.floor(posWorld.x - widthMap),
        posIntY = Math.floor(posWorld.y - heightMap),
        posEndX = Math.floor(posWorld.x + widthMap),
        posEndY = Math.floor(posWorld.y + heightMap);

    if ((posIntX >= this.pos.x || this.pos.x <= posEndX) && (posIntY >= this.pos.y || this.pos.y <= posEndY)) {
        for(let y = 0, intY = posIntY; intY <= posEndY; intY++, y++) {
            for(let x = 0, intX = posIntX; intX <= posEndX; intX++, x++) {

                if (this.pos.x == intX && this.pos.y == intY) {
                    return {x: x, y: y};
                }
            }
        }
    }

    return false;
}

/* ------------------------------ *
    DRAW
* ------------------------------ */

draw (ctx, cXnull, cYnull) {
    const tileSize = 32;
    cXnull *= tileSize;
    cYnull *= tileSize;

    if (this.moving && (Date.now() - this.lastFrameUpdate > 150)) {
        console.log(`Moviendo ${Date.now()} - ${this.lastFrameUpdate}`);
        this.lastFrameUpdate = Date.now();
        this.nextFrame();
    } else if (!this.moving) {
        this.frame = 0;
    }

    //console.log("Draw player"+ this.dir);
    ctx.fillStyle = "#FFF";
    ctx.font = "9pt Minecraftia";
    //ctx.fillText(this.playerName, this.absPos["absX"] + cXnull, this.absPos["absY"] + cYnull-10);
    //nuevo
    ctx.fillText(this.name +', '+ this.level, cXnull, cYnull - 32);
    ctx.fillStyle="#000000";
    ctx.fillRect(cXnull, cYnull - 30, 50, 2);
    ctx.fillStyle="#FF0000";
    ctx.fillRect(cXnull, cYnull - 30, 40, 2);
    if (this.mode == 0) {
        //nuevo
        ctx.drawImage(this.skinBase, this.frame * 64, this.dir * 48, 64, 48, (cXnull - 16), cYnull, 64, 48);

    } else if (this.mode == 1) {
        ctx.drawImage(this.skinBase, this.fightFrame*44, this.mode*44, 44, 44, this.absPos["absX"] + cXnull-5, this.absPos["absY"] + cYnull-5, 32, 32);
    }
    //ctx.drawImage(this.playersprite, this.frame*42, this.dir*43, 42, 43, this.absPos["absX"] + cXnull, this.absPos["absY"] + cYnull, 32, 32);

    let hitDelta = Date.now() - this.lastHitPointUpdate;

    if (this.hitArray.length > 0 && hitDelta > 50) {
        for (var i = 0, j = this.hitArray.length; i < j; i+=1) {
            if (this.hitArray[i] != null) {
                this.lastHitPointUpdate = Date.now();
                this.hitArray[i][1] = Math.round((this.hitArray[i][1]-0.1)*10)/10;

                if (this.hitArray[i][1] <= 0) {
                    this.hitArray.splice(i,1);
                }
            }
        }
    } else if (this.hitArray.length == 0) {
        this.gotHit = false;
    }

    if (this.hitArray.length > 0) {
        for (var i = 0, max = this.hitArray.length; i < max; i += 1) {
            ctx.fillStyle = 'rgba(255,0,0,'+this.hitArray[i][1]+')';
            ctx.font = "14pt Minecraftia";
            ctx.fillText(this.hitArray[i][0], this.pos["x"]+12, this.pos["y"]-20);
        }
    }
}

}