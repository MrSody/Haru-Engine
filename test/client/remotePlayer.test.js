/**
 * @jest-environment jsdom
 */

const CONSTANT = require('./constants');
const { createCanvas, loadImage } = require('canvas');

import { response } from 'express';
import RemotePlayer from '../../client/public/game/js/modules/entities/player/remotePlayer';

test('posNow', () => {
    let player = new RemotePlayer(CONSTANT.DATA_PLAYER);
    let widthMap = 100;
    let heightMap = 100;
    let posWorld = {x: 10, y: 5};

    expect(player.posNow(widthMap, heightMap, posWorld)).toStrictEqual({x: 99, y: 106});
});

test('draw', async () => {
    const player = new RemotePlayer(CONSTANT.DATA_PLAYER);
    const coordinateX = 5;
    const coordinateY = 5;
    const tileSize = 32;

    const canvaCharacters = createCanvas(544, 352);
    const ctxCharacters = canvaCharacters.getContext('2d');
    const canvaHUB = createCanvas(544, 352);
    const ctxHUB = canvaHUB.getContext('2d');

    const image = await loadImage(`${__dirname}\\spritePlayer.png`);

    player.skinBase = image;

    player.draw(ctxCharacters, ctxHUB, coordinateX, coordinateY, tileSize);

    const dataURLCharacters = canvaCharacters.toDataURL();
    expect(dataURLCharacters).toMatchSnapshot();

    const dataURLHUB = canvaHUB.toDataURL();
    expect(dataURLHUB).toMatchSnapshot();
});