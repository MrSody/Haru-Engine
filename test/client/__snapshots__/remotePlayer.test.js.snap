// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`draw 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAiAAAAFgCAYAAACCOcLtAAAABmJLR0QA/wD/AP+gvaeTAAAJDklEQVR4nO3dXWjdZx0H8G/OaefpXp0b66ZO546TMabEMVxQNuKF4wyEKaJGvVhAdF4IFhzaC9GAEyKO0V1MsLOaK3PmhWz0YrkZFHphJjLLRDamkRas3cZcdWtnui6JF/FkeW2ak3NynpTPB0pycv7/33luEr59Xn7/BAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAYKm+Xg+A7W0qM3Nb8Tnj2Z/RfC+n8sZ649mK4Syop+p3CKANlV4PANYzkd/lR/n2uuEjSeqp5jf5xRaMCoDN8L83NqXbMyCv5EQ+k1tyKq9v6L4f5+f5au7v0qjeYQYEoD1mQCjaL/PwhsNHkjTzWBdGA0CnCCAU65WcyMGMt3XvX/InSzEABRNAKNaz+X1On8e+DwC2HwGEou3MRb0eAgBdIIBQrHfnPXl/bmj7/q3YhApAewQQijWQwdyRwV4PA4AuEEAo2oE8nJuvv3ZD91QrfZnctzf1VLs0KgA2SwCheI88+P1cffml53VttdKXB75wd/objS6PCoDN2NHrAcC5TGUm9fuq+e69t+Xw86/lDy8eXfPa3bt3Z/iu/uyq/ie1mwe3vC07AOfPDAjbwhc/fXd++LXP5uDBg/nc3XflyisuS+1dF+XiXbV88raP5icP3J9jx47l85/6eBoDg70eLgDrMANC0Vr7OCYmD6UxMJj+my7LrT/4Tl49+c+k8mauueqKXHv19ZmbvSQ5OrlwbetesyAAZRJAKFY91YwMDSRJRpqTufUT0/nwG3/OB2+czQ2Vp5ZcOzd7T147Uck/Zo9npHkkk/v2ztfYI4QAlMiDtNiUbj2Mrp7qQoioXX4m19SP57qbfnte957465dy1XX3ZW72khyZmMjAntGuhRAPowNojz+ebEo3AkgrfPQ3GumrnM5bZ0dz6ZWHN1Rjduabefu/93Y9hAggAO2xCZWiLA8fO3Y9ueHwkSSV6v7s2PVk+iqn099o6AsCUBgBhGIsDh9JUt35fCrV/W3Xq1T3p7rz+U4ND4AOMn3MpnRyCaaeaqZfOJQkqex4OTsv/nJH6p598/HMvr27K71BLMEAtMcMCEVYvOk0SSo7n+tY7VYtyzAA5RBAKMbi9umV6rMdq9uqpT07QDkEEIrTVzmdvmV9PjZX76n0VU53rB4AmyeA0HPLl1/6Kqc6/hndqAlA+wQQirBVyyP2gQCUQQChKKMje/Orx37W8bpzs5cmsQ8EoBQCCD3VWn45MjGx8LNv/fSPSXK2U5/x1vQt8w+rA6AYAghFWGVm4minar/xrw91qhQAHSKA0HNrLIs82qn6rx79QKdKAdAhAghF6ntvHnl644+AWeHpw8m/j1+x8Lob3VAB2DgBhJ6aykxqNw8mmd+Autj4E8nLL13Xdu1TJ+/M+BM2ngKUSAChGHtHRtMYGMzI0ECSzB0YT554/H05dfLODdc6dfLO7HnwcL5xx971LwZgywkgFKW/0UhjYHDh9UvHdmTqmdszc+ah864xc+ahTD1zew6ML539sPwCUI4dvR4ALDewZ3Th+5HmZEaak0mSt/72eCo7n0ul+uyKVu1zs/dkdua2zJ79WGbf3p3p11/d0jEDsDECCEVp9QOZykzqqWZsbCxDQ0NJkotqtYXrvv6V5CM3zn//4t+TA+NPJVkaSha3dwegLAIIPTe/EbWa6RcOrXivFT6SZHp6es0aj/56/muz2UySDA8P23wKUDB7QChC6zTMwJ7RdfdptELGalqBZWxsbOF0TZIlnVYB6D0BhGJMZabtTaKLQ8niWZNkPnycT7ABYOtYgmHbGRoaSrPZXAgaa82ILJ4FET4AymIGhG2rVqulVqtleHh4xXuLl2IAKI8ZEIpWq9VWbD5tNpsZHh5eMqtRH66uuHe1YAJAGQQQija5b29qi47ftqy2pLI8cPy/o+pCHxEAyiGAULT+RiOT+955PTF5aM1A0QocLa2OqgIIQHkEEIq3uJ/HuU6zNAYG1+j9MbrKzwDoJZtQ2Tba6eWh/wdAmcyAsC0cmZjIxOShc17Ter+/0VgIHvp/AJSpr9cDYHubysxct2rXs7I9+1pPtK1n5SmYlm4GkHqqfocA2mAGhAvGaj0/HMUFKJMAwgVhKjOpD1eX9Ayp1WqWXwAKJYBQtMUPlEvmZznqw9VVg8VUZlKrVZe8BqBMTsFQpHqqGRsby/T09JJ/62k90E74ACibAEJxWuFj+VNtAbhwCCAU5VzhY7VnwACwPTlCyKZ0+hhuPdVVl1pKDR+O4QK0xyZUirJ8I+ny9wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIALyP8AF+/lMgX3xN8AAAAASUVORK5CYII="`;

exports[`draw 2`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAiAAAAFgCAYAAACCOcLtAAAABmJLR0QA/wD/AP+gvaeTAAAHmUlEQVR4nO3ca4inVR0H8O8p81JsmhcUSZYughqEQZaRICik1BtjyURkq00rJBQslQqksAyKwIy0QBO1IC0qk/LSC4mwsBST1BQMKi+Vt7TU9ZbfXjzPyjDO6qyXZtg+n1d/zjnP7/zOw/yZM+d3mAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABWhbY7tH10pfNYStsj2168jHGntH2o7TdfhhxubPvWlzouwHK8YqUTAJbWdr8k65OsTfJU2xNWOCWAl8w2K50AsFn7J7lujPFgkhNXOhkA2Gp0cnzbW9o+2Pb7bXdsu77tBW3XtT2v7WvaPrzo2Wfa2u7R9jttz297adtPzSWYjfPn2+b4F7VdMz/zgbb3tX2s7dVtd15Gvpe3PXyJtiPmOOvb/r7tI22vabvvPGa3ee4H2t7T9tpNJZi2h7a9pO2FbX/W9oC2x/bZTpjH/6HtJ9v+su2/297a9rC5b/u2X53fxcVtz2m7/dy3bn720bbXz2tXggFWhBIMq8FuSd6bZM8ktyT50tz+ziS3jjE++lwPtx1JLkly9hjjI0mOSHJokkOSbJtkTZL3JNkryT1JTkuSMcYPxhi7JtklycYkH3uR63hVplOLo+Y1XZrk7Dm/c5N8a17j3kmumHPfJ8kXk3xojLE+yceTXJjkoiTHJTl3jDGSnLVorkOSnJJk9yQnJPle222TfDnJn8YYG8YYH0zyjyQntT0oyfuSHJbktUmOTvJwAFaIEgwrbozxhU2f256T5Pokv01y7Rjj5mWE2D/J02OM6+Z4bXtZksOTPD7G+PyC+N9IcmWSkxfM/0jbK5Ls+yKX8sQY46QFc52X5LNJ9kmy0xjjmrnrsbY3z+0bkpw/xtg453Jn23uXkctpY4yb5s9Xtf1XkjckOSbJHgvGXZrkzCRvSXLBGOPuuf22tg+90IUCvFg2IKw2DyXZdQufWZvk4LZd1P6jJcb+c1P8tsdlOq3YL9OJxblbOO/zeSTJdknelOSBzYxZm+Tktt9e1L7nC5hrr0xre2rRq7g900nQ/VsYE+BlowTDarNLkjuWaH8yybZtl/qZvSPJ5WORTKcBi+2W5K9tD81UwvhKkrdlKmMsx+NJdljm2E3+nqnssZQ7k6xfnPsY4+dbOEeS3Jvkb0vE2jtTKWZzOQD8z9mAsOLavqPtdm33SHJkkqsWjxljPJHkj0k+3PbV8+XJM+fuG5OsaXvQ3Lem7VFJTk3yyrZvn+PvmulexU+S7JjkriQ3ZNr0HLkgn+3mi5rrlkj3hiRHt31d29e3PTHJgc+zxBuT7N72wPli7PuTfGbuuyTJurZvbLtN27VtT2/7rud/c8/ynyS/aHtM253mdRzc6X+I/DjJsW13brtn27MyncwArAgbEFaD0zOVKK7LVJI4dTPjNiQ5PtNf+ucn+XWSjDGeyrSBOC7JbUnuTvLuTBc3n850qfX+THdLNiY5I8llc5w/Z7r0efuCeQ7IVAr66RI5fC3TL/q/JPlVpvsWNy0x7hljjCczncZ8fV7nyUl+OPddO7dfmOS+JFcnuXmM8ZvnivkcPpHpbsnv5rk+nWm938201lvmfO/K0idNALD1W+Lexopr+7m2Z6x0HgDAy2SVbkCubPvmlc4DYGs2VjoB/r81WXUbkIW24AviuwSwBdwBAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAALYq/wWtjYiJm4jItgAAAABJRU5ErkJggg=="`;
